### ISSUE TITLE
Migration to 1.4 fails, if there are too many zk entries

### ISSUE THREAD

**Issue by aquamatthias** [mesosphere/marathon#5083]
The Migration fails with this exception:
```
marathon [2017-02-02 13:56:56,421] ERROR Failed to take over leadership (mesosphere.marathon.core.election.impl.CuratorElectionService:pool-4-thread-1)
marathon mesosphere.marathon.MigrationFailedException: while migrating storage to 1.2
marathon 	at mesosphere.marathon.storage.migration.Migration$$anonfun$migrations$6$$anonfun$apply$5.applyOrElse(Migration.scala:96)
marathon 	at mesosphere.marathon.storage.migration.Migration$$anonfun$migrations$6$$anonfun$apply$5.applyOrElse(Migration.scala:95)
marathon 	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
marathon 	at scala.util.Failure$$anonfun$recover$1.apply(Try.scala:216)
marathon 	at scala.util.Try$.apply(Try.scala:192)
marathon 	at scala.util.Failure.recover(Try.scala:216)
marathon 	at scala.concurrent.Future$$anonfun$recover$1.apply(Future.scala:326)
marathon 	at scala.concurrent.Future$$anonfun$recover$1.apply(Future.scala:326)
marathon 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
marathon 	at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:121)
marathon 	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
marathon 	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
marathon 	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
marathon 	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
marathon Caused by: mesosphere.marathon.StoreCommandFailedException: Could not fetch MarathonTaskState with key: restservice.0f5e195a-de00-11e6-b397-560ba27a4ebd
marathon 	at mesosphere.marathon.storage.repository.legacy.store.MarathonStore$$anonfun$fetch$1.applyOrElse(MarathonStore.scala:46)
marathon 	at mesosphere.marathon.storage.repository.legacy.store.MarathonStore$$anonfun$fetch$1.applyOrElse(MarathonStore.scala:41)
marathon 	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
marathon 	at scala.util.Failure$$anonfun$recover$1.apply(Try.scala:216)
marathon 	at scala.util.Try$.apply(Try.scala:192)
marathon 	at scala.util.Failure.recover(Try.scala:216)
marathon 	at scala.concurrent.Future$$anonfun$recover$1.apply(Future.scala:326)
marathon 	at scala.concurrent.Future$$anonfun$recover$1.apply(Future.scala:326)
marathon 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
marathon 	at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:121)
marathon 	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
marathon 	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.pollAndExecAll(ForkJoinPool.java:1253)
marathon 	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1346)
marathon 	... 2 common frames omitted
marathon Caused by: java.lang.IllegalStateException: ZkStore queue may not exceed 1024 entries
marathon 	at mesosphere.marathon.util.WorkQueue.apply(WorkQueue.scala:72)
marathon 	at mesosphere.marathon.storage.repository.legacy.store.ZKStore.load(ZKStore.scala:38)
marathon 	at mesosphere.marathon.storage.repository.legacy.store.MarathonStore.fetch(MarathonStore.scala:34)
marathon 	at mesosphere.marathon.storage.repository.legacy.store.EntityStoreCache$$anonfun$fetch$1.apply(EntityStoreCache.scala:32)
marathon 	at mesosphere.marathon.storage.repository.legacy.store.EntityStoreCache$$anonfun$fetch$1.apply(EntityStoreCache.scala:32)
marathon 	at mesosphere.marathon.storage.repository.legacy.store.EntityStoreCache.directOrCached(EntityStoreCache.scala:121)
marathon 	at mesosphere.marathon.storage.repository.legacy.store.EntityStoreCache.fetch(EntityStoreCache.scala:32)
marathon 	at mesosphere.marathon.storage.repository.legacy.LegacyEntityRepository$stateMachine$macro$1238$1$$anonfun$5.apply(LegacyEntityRepository.scala:51)
marathon 	at mesosphere.marathon.storage.repository.legacy.LegacyEntityRepository$stateMachine$macro$1238$1$$anonfun$5.apply(LegacyEntityRepository.scala:51)
marathon 	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
marathon 	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
marathon 	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
marathon 	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
marathon 	at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
marathon 	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
marathon 	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
marathon 	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
marathon 	at mesosphere.marathon.storage.repository.legacy.LegacyEntityRepository$stateMachine$macro$1238$1.apply(LegacyEntityRepository.scala:51)
marathon 	at mesosphere.marathon.storage.repository.legacy.LegacyEntityRepository$stateMachine$macro$1238$1.apply(LegacyEntityRepository.scala:45)
marathon 	... 7 common frames omitted
```

**Comment by aquamatthias** [comment]
@jasongilanfarr can you have a look?

**Comment by jdef** [comment]
I remember talking about the queue length with jason; decision (at the time) was to generate an error if the queue length was exceeded because that's what the old code did. an alternative would be to block until the queue length falls below some threshold (e.g. max, or max/2).

**Comment by jdef** [comment]
https://phabricator.mesosphere.com/D500 ?

**Comment by jasongilanfarr** [comment]
D500 (probably with Int.MaxValue) instead should be sufficient. The proper solution is in #5152

**Comment by jdef** [comment]
landed directly into releases/1.4; no cherry-pick needed there

filed a separate ticket for forward-porting the change to master
